src/blockencodings.cpp

```diff
@@ -75,12 +76,20 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
     prefilled_count = cmpctblock.prefilledtxn.size();
 
     // Calculate map of txids -> positions and check mempool to see what we have (or dont)
-    std::map<uint64_t, uint16_t> shorttxids;
+    // Because well-formed cmpctblock messages will have a (relatively) uniform distribution
+    // of short IDs, any highly-uneven distribution of elements can be safely treated as a
+    // READ_STATUS_FAILED.
+    std::unordered_map<uint64_t, uint16_t> shorttxids(cmpctblock.shorttxids.size());
     uint16_t index_offset = 0;
     for (size_t i = 0; i < cmpctblock.shorttxids.size(); i++) {
         while (txn_available[i + index_offset])
             index_offset++;
         shorttxids[cmpctblock.shorttxids[i]] = i + index_offset;
+        // The math for a max bucket size of 10 is rather complicated, but simulation
+        // shows that, for blocks of up to 10k transactions, any individual bucket having
+        // more than 10 entries is highly unlikely.
```

Comment:
How accurate was your simulation? If I approximate the chance as buckets*P[binomial(n,1.0/buckets)], where n<=buckets (due to max_load_factor=1.0 by default for unordered_map), this would randomly fail about once every 9000 transmissions for n=10000 buckets=10000.

According to this approximation, if we want the random failure chance to be less than once in a billion, you need:
* `> 10` for up to 13 transactions
* `> 11` for up to 23 transactions
* `> 12` for up to 56 transactions
* `> 13` for up to 291 transactions
* `> 14` for up to 3423 transactions
* `> 15` for up to 53645 transactions
* `> 16` for up to 913429 transactions